modules/library/metadata/src/main/java/org/geotools/factory/FactoryRegistry.java

```diff
@@ -152,15 +158,52 @@ public FactoryRegistry(final Class<?>[] categories) {
      * @param categories The categories.
      */
     public FactoryRegistry(final Collection<Class<?>> categories) {
-        super(categories.iterator());
-        for (final Iterator<Class<?>> it=getCategories(); it.hasNext();) {
-            if (needScanForPlugins == null) {
-                needScanForPlugins = new HashSet<Class<?>>();
-            }
-            needScanForPlugins.add(it.next());
+        for (Class<?> category : categories) {
+            needScanForPlugins.add(category);
+            registry.registerCategory(category);
         }
     }
 
+    /*
+     * TODO: discuss how much API can be broken
+     *
+     * Breakage:
+     *
+     *  - not all methods of `ServiceRegistry` API were implemented
+     *
+     * Possible breakage:
+     *
+     *  - in general nomenclature, replace "ServiceProvider" with "Factory"
```

Comment:
The strategy for that may require some more discussion, but deprecations with the functionality still available, particularly with the goal of fitting with the naming conventions used by the rest of the GeoTools API wouldn't worry me.